pkg/agent/route/route.go

```diff
@@ -0,0 +1,293 @@
+
+	if c.nodeConfig.TrafficEncapMode.SupportsNoEncap() {
+		ServiceRtTable.Idx = AntreaServiceTableIdx
+		ServiceRtTable.Name = AntreaServiceTable
+	}
+
+	if ServiceRtTable.IsMainTable() {
+		c.RemoveServiceRouting()
+		return nil
+	}
+	f, err := os.OpenFile("/etc/iproute2/rt_tables", os.O_RDWR|os.O_APPEND, 0)
+	if err != nil {
+		klog.Fatalf("Unable to create service route table(open), err %v", err)
+	}
+	defer func() { _ = f.Close() }()
+
+	oldTablesRaw := make([]byte, 1024)
+	bLen, err := f.Read(oldTablesRaw)
+	if err != nil {
+		klog.Fatalf("Unable to create service route table(read), err %v", err)
+	}
+	oldTables := string(oldTablesRaw[:bLen])
+	newTable := fmt.Sprintf("%d %s", ServiceRtTable.Idx, ServiceRtTable.Name)
+
+	if strings.Index(oldTables, newTable) == -1 {
+		if _, err := f.WriteString(newTable); err != nil {
+			klog.Fatalf("Failed to add antrea service route table, err=%v", err)
+		}
+	}
+
+	gwConfig := c.nodeConfig.GatewayConfig
+	if gwConfig != nil && c.nodeConfig.PodCIDR != nil {
+		// add local podCIDR if applicable to service rt table
+		route := &netlink.Route{
+			LinkIndex: util.GetNetLink(gwConfig.Link).Attrs().Index,
+			Scope:     netlink.SCOPE_LINK,
+			Dst:       c.nodeConfig.PodCIDR,
+			Table:     ServiceRtTable.Idx,
+		}
+		if err := netlink.RouteReplace(route); err != nil {
+			klog.Fatalf("Failed to add link route to service table, err=%v", err)
+		}
+	}
+
+	// create ip rule to select route table
+	ipRule := netlink.NewRule()
+	ipRule.IifName = util.GetNetLink(c.nodeConfig.GatewayConfig.Link).Attrs().Name
+	ipRule.Mark = iptables.RtTblSelectorValue
+	ipRule.Table = ServiceRtTable.Idx
+
+	err = netlink.RuleAdd(ipRule)
+	if err != nil {
+		if os.IsExist(err) {
+			klog.Errorf("rule already created")
+		} else {
+			klog.Fatalf("Failed to create ip rule for service route table, err=%v", err)
+		}
+	}
+	return nil
+}
+
+// AddPeerCIDRRoute adds routes to route tables for Antrea use.
+func (c *Client) AddPeerCIDRRoute(peerPodCIDR *net.IPNet, gwLinkIdx int, peerNodeIP, peerGwIP net.IP) ([]*netlink.Route, error) {
+	if peerPodCIDR == nil {
+		klog.Errorf("Empty peer pod CIDR")
+		return nil, nil
+	}
+
+	// install routes
+	routes := []*netlink.Route{
+		{
+			Dst:       peerPodCIDR,
+			Flags:     int(netlink.FLAG_ONLINK),
+			LinkIndex: gwLinkIdx,
+			Gw:        peerGwIP,
+			Table:     ServiceRtTable.Idx,
+		},
+	}
+
+	// If service route table and main route table is not the same , add
+	// peer CIDR to main route table too (i.e in NoEncap and hybrid mode)
+	if !ServiceRtTable.IsMainTable() {
+		if c.nodeConfig.TrafficEncapMode.NeedEncap(peerNodeIP, c.nodeConfig.NodeIPAddr) {
+			// need overlay tunnel
+			routes = append(routes, &netlink.Route{
+				Dst:       peerPodCIDR,
+				Flags:     int(netlink.FLAG_ONLINK),
+				LinkIndex: gwLinkIdx,
+				Gw:        peerGwIP,
+			})
+		} else {
+			routes = append(routes, &netlink.Route{
+				Dst:       peerPodCIDR,
+				Flags:     int(netlink.FLAG_ONLINK),
```

Comment:
Question - why we need the ONLINK flag and specify the dev? For hybrid mode, I guess we need not this. So, maybe you mean for noEncap mode, you want to force the packet to be sent to the destination Node IP, assuming source Node can resolve the destination Node's MAC, even it is in a different subnet?
Do you really see such scenarios (Nodes can reach each other even they are not in the same subnet)? I was thinking if we really want to support NoEncap mode (besides hybrid. You know I do not really suggest to add NoEncap which is not very useful), we might assume someone or some code configured underlay router to route Pod traffic across different Node subnets, and in this case we should not add routes in the source Node. What you think?

Reply:
yes, the onlink flag is to be safe to noEncap mode where peer IP is not on the same subnet but is on the same physical network.

Reply:
I hope we keep to keep NoEncap mode as it provide additional flexibility to our user.

Reply:
K8s Nodes are not really that ephemeral - they are VMs not Pods, and needs to install K8s and join the Cluster, typically add/removed manually (I mean manually triggered).
Even you believe manual route configuration is not common, do you think there can another controller to automate router configuration at Node creation/removal? How about your DHCP idea?

But do you see real deployments of using multiple subnets in a single physical L2/VLAN? I guess it is real.

Again, as I said earlier, in general I do not think NoEncap is very useful in either of the two cases.

Reply:
Of course, one possible route automation is by BGP.

Reply:
1. regarding to how worker node IP can be changed. My experience is that user does not need to remove and add node to k8s cluster. An power off/on of worker node is suffice to change its node ip, assuming kubelet is started with --node-ip DHCLIENT_IP. 

2.regarding to the usefulness to NoEncap mode, my thinking is that we don't know what users deployment case looks like, why not give that option? And we know in circumstances that multiple IP subnets may be accessible on the same L2 network. I personally do not believe that hybrid covers all potential use cases,  that is reason it make sense to me that we give the user to option to configure NoEncap and Encap in addition to hybrid.


But I diverse, the question is that in NoEncap mode, if peerNode is not on the same subnet, should peerPodCIDR nexthop be peerNode(IP) with onlink (this works if all worker nodes are reachable via L2); or nexthop be default gw (this works if  underlay router knows how to route pod CIDR traffic)?

Reply:
Yes. I agree with your last summary. So, we have three options:
1. Do not support NoEncap at all, if no real use case. It will also be hard to test.
2. Define it as all Nodes in the same physical L2, even they are not in the same subnet - your current implementation.
3. Define it as Node route table and underlay router will handle the routing, which can have two cases: 1) user manually configure routes; 2) there is another solution automatically configures routes or uses BGP to advertise routes. Last how if the Pod IPs are allocated by DHCP - it should be treated as noEncap or passthrough?

I do not quite like #2 because I assume most people will think about #3 when talking about noEncap, esp. with other solutions like Calico.
If you like, we can have more offline discussion.

Reply:
renamed noEncap to noEncapSameL2 as discussed offline